test/testdrive/subexpression-replacement.td

```diff
@@ -0,0 +1,120 @@
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+#
+# Tests for the optimization described in https://github.com/MaterializeInc/materialize/pull/6196/
+# Additional tests in test/sqllogictest/transform/predicate_reduction.slt
+#
+
+$ set-sql-timeout duration=125ms
+
+# Remove both newlines, references to internal table identifiers, and "materialize.public" strings, all with a single regexp
+$ set-regex match=(\s\(u\d+\)|\n|materialize\.public\.) replacement=
+
+> CREATE TABLE t1 (col_null INTEGER, col_not_null INTEGER NOT NULL);
+
+> INSERT INTO t1 VALUES (1, 1);
+
+# The simplest expression there could be
+
+> EXPLAIN SELECT * FROM t1 WHERE col_null IS NULL AND (col_null IS NULL AND col_not_null = 5);
+"%0 =| Get t1| Filter isnull(#0), (#1 = 5)"
+
+> EXPLAIN SELECT * FROM t1 WHERE col_not_null = 1 AND (col_not_null = 1 AND col_null = 5);
+"%0 =| Get t1| Filter (#0 = 5), (#1 = 1)"
+
+# NULL-able expressions are not optimized
+> EXPLAIN SELECT * FROM t1 WHERE col_null = 1 AND (col_null = 1 AND col_not_null = 5);
+"%0 =| Get t1| Filter (#0 = 1), (#0 = 1), (#1 = 5)"
+
+# OR/disjunction at the top level
+
+> EXPLAIN SELECT * FROM t1 WHERE col_null IS NULL OR (col_null IS NULL AND col_not_null = 5);
+"%0 =| Get t1| Filter isnull(#0)"
+
+> EXPLAIN SELECT * FROM t1 WHERE col_null IS NULL OR col_null IS NULL OR (col_null IS NULL AND col_not_null = 5);
+"%0 =| Get t1| Filter isnull(#0)"
+
+> EXPLAIN SELECT * FROM t1 WHERE col_null IS NULL OR (col_null IS NULL AND col_not_null = 5) OR (col_null IS NULL AND col_not_null = 6);
+"%0 =| Get t1| Filter isnull(#0)"
+
+# OR/disjunction at the lower level
+
+> EXPLAIN SELECT * FROM t1 WHERE col_null IS NULL AND (col_null IS NULL OR col_not_null = 5);
+"%0 =| Get t1| Filter isnull(#0)"
+
+# Nested OR/disjunctions are not optimized
+
+> EXPLAIN SELECT * FROM t1 WHERE col_null IS NULL OR (col_null IS NULL OR col_not_null = 5);
+"%0 =| Get t1| Filter (isnull(#0) || (isnull(#0) || (#1 = 5)))"
```

Comment:
This is not optimized I believe because nested-ORs are not flattened. I will open a separate ticket about flattening the ORs.